File: Example Code1/RadioHub/RadioHub.py
'''
 @file    RadioHub.py
 
 @brief   Implementation of class RadioHub as main interface to further service access points like "DeviceManagement"
 
 @note    This example code is free software: you can redistribute it and/or modify it.
 
          This program is provided by IMST GmbH on an "AS IS" basis without
          any warranties in the hope that it will be useful.
 '''
 
import serial
import sliplib.slip

from RadioHub.ServiceAccessPoints.ServiceAccessPoint import ServiceAccessPoint_ID
from RadioHub.ServiceAccessPoints.DeviceManagement import DeviceManagement
from RadioHub.ServiceAccessPoints.Trace import Trace
from RadioHub.Utils.SerialMessage import SerialMessage

class RadioHub:

    def __init__( self ):
        self.serialDevice = None
        self.slip = None
        self.deviceManagement = None
        self.serviceAccessPoints = {}

    def open( self, portName, client ) -> bool:

        # save client    
        self.client = client

         # open serial port
        self.serialDevice = serial.Serial( port = portName, baudrate = 115200, timeout = 0 )
        
        # create SLIP encoder/decoder
        self.slip = sliplib.slip.Driver()

        # create DeviceManagement SAP
        self.deviceManagement  = DeviceManagement( client = client, serialDevice = self.serialDevice, sap_id = ServiceAccessPoint_ID.DeviceMgmt ) 
        # and add to dictionary
        self.serviceAccessPoints[ ServiceAccessPoint_ID.DeviceMgmt ] = self.deviceManagement

        # create and add Trace SAP
        self.serviceAccessPoints[ ServiceAccessPoint_ID.Trace ] = Trace( client = client, serialDevice = self.serialDevice, sap_id = ServiceAccessPoint_ID.Trace )
        
        return True

    # perform message receiver tasks
    def process( self ):

        # read data from serial device    
        data = self.serialDevice.read( 100 )

        if ( len( data ) > 0 ):
            
            # get slip decoded messages
            messages = self.slip.receive( data )        
            
            # decode HCI messages
            for msg in messages:
                serialMsg = SerialMessage()

                # decde byte stream and check crc
                if ( serialMsg.decode( msg ) == True ):
                    # dispatch HCI message
                    result = {}
                    try:
                        # find sap
                        sap = self.serviceAccessPoints[ serialMsg.sap_id ]
                    except:
                        # error
                        result[ "error" ] = "unknown service access point called: " + hex( serialMsg.sap_id )     
                    else:
                        # call message decoder
                        sap.decodeMessage( serialMsg = serialMsg, result = result )
                    
                    # notify client about results    
                    self.client( result )

File: Example Code1/RadioHub/RadioHub.cpp
/**
 * @file    RadioHub.cpp
 *
 * @brief   Implementation of class RadioHub
 *
 * @note    This example code is free software: you can redistribute it and/or modify it.
 *
 *          This program is provided by IMST GmbH on an "AS IS" basis without
 *          any warranties in the hope that it will be useful.
 */

#include "RadioHub.h"
#include <QSerialPortInfo>

/**
 * @brief   class constructor
 *
 * @param   client      reference to client
 */
RadioHub::RadioHub      ( RadioHub::Client& client )
        : _Client       ( client )
        , _DeviceMgmt   ( _Port )
        , _SlipDecoder  ( this )
{
    // connect to serial port for ready read events
    connect( &_Port, SIGNAL( readyRead() ), this, SLOT( OnSerialPort_ReadyRead() ) );
}

/**
 * @brief   enable RadioHub for communication with connected radio module
 *
 * @param   portName    name of serial port, e.g.: "ttyUSB", "COM5"
 *
 * @return  true/false
 */
bool
RadioHub::Enable( const QString& portName )
{
    // open serial port
    if ( _Port.isOpen() == true )
        return false;

    // reset SLIP decoder
    _SlipDecoder.Reset();

    // clear serial message buffer
    _RxMessage.clear();

    // configure port name
    _Port.setPortName( portName );

    // open port for read and write access
    if ( _Port.open( QIODevice::ReadWrite ) == false )
        return false;

    _Port.setBaudRate( QSerialPort::Baud115200 );
    _Port.setDataBits( QSerialPort::Data8 );
    _Port.setStopBits( QSerialPort::OneStop );
    _Port.setParity( QSerialPort::NoParity );
    _Port.setFlowControl( QSerialPort::NoFlowControl );

    QJsonObject info;
    info[ "Port" ]      = _Port.portName();
    info[ "Baudrate" ]  = QString::number( QSerialPort::Baud115200 ) + " bps";

    QSerialPortInfo pi = QSerialPortInfo( _Port );
    info[ "Manufacturer" ] = pi.manufacturer();

    // init object
    _DeviceInfo[ "Connection" ] = info;

    return true;
}

/**
 * @brief   process available bytes from serial port
 */
void
RadioHub::OnSerialPort_ReadyRead()
{
    // get number of received bytes
    quint64 numBytes =  _Port.bytesAvailable();

    if ( numBytes > 0 )
    {
        // pass incoming byte stream to SLIP decoder
        _SlipDecoder.Decode( _RxMessage, _Port.readAll() );
    }
}

/**
 * @brief   process SLIP decoded HCI messages
 */
void
RadioHub::OnSlipDecoder_MessageReady( const QByteArray& /* msg */ )
{
    // HCI message is aligned in _RxMessage, we can ignore incoming param "msg" here

    QJsonObject result;

    // pass message to message decoder and convert message content into human readable JsonObject
    if ( ServiceAccessPoint::OnDispatchMessage( _RxMessage, result ) )
    {
        _Client.OnRadioHub_DataEvent( result );
    }
}
